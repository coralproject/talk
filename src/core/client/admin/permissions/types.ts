import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
