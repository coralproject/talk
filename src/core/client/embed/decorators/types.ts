import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
