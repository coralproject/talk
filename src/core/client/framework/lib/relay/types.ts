import { _RefType } from "react-relay";

export type FragmentKeys<T> = {
  [P in keyof T]: T[P] extends _RefType<any> | null ? P : never
}[keyof T];

export type FragmentKeysNoLocal<T> = Exclude<FragmentKeys<T>, "local">;
