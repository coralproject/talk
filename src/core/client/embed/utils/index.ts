export { default as buildURL } from "./buildURL";
export { default as ensureEndSlash } from "./ensureEndSlash";
export { default as ensureNoEndSlash } from "./ensureNoEndSlash";
export { default as startsWith } from "./startsWith";
export { default as prefixStorage } from "./prefixStorage";
export { default as parseHashQuery } from "./parseHashQuery";
