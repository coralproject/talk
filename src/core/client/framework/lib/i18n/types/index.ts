export { default as FluentShortNumber } from "./FluentShortNumber";
