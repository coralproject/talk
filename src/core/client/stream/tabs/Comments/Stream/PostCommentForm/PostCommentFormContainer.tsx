import { Localized } from "@fluent/react/compat";
import { FORM_ERROR } from "final-form";
import React, { Component } from "react";

import { withContext } from "coral-framework/lib/bootstrap";
import {
  InvalidRequestError,
  ModerationNudgeError,
} from "coral-framework/lib/errors";
import {
  FetchProp,
  graphql,
  withFetch,
  withFragmentContainer,
} from "coral-framework/lib/relay";
import { PromisifiedStorage } from "coral-framework/lib/storage";
import { PropTypesOf } from "coral-framework/types";
import {
  ShowAuthPopupMutation,
  withShowAuthPopupMutation,
} from "coral-stream/mutations";

import { PostCommentFormContainer_settings } from "coral-stream/__generated__/PostCommentFormContainer_settings.graphql";
import { PostCommentFormContainer_story } from "coral-stream/__generated__/PostCommentFormContainer_story.graphql";
import { PostCommentFormContainer_viewer } from "coral-stream/__generated__/PostCommentFormContainer_viewer.graphql";
import {
  COMMENT_SORT,
  COMMENTS_TAB,
} from "coral-stream/__generated__/StreamContainerLocal.graphql";

import {
  getSubmitStatus,
  shouldTriggerSettingsRefresh,
  SubmitStatus,
} from "../../helpers";
import RefreshSettingsFetch from "../../RefreshSettingsFetch";
import {
  CreateCommentMutation,
  withCreateCommentMutation,
} from "./CreateCommentMutation";
import PostCommentForm from "./PostCommentForm";
import PostCommentFormClosed from "./PostCommentFormClosed";
import PostCommentFormClosedSitewide from "./PostCommentFormClosedSitewide";
import PostCommentFormFake from "./PostCommentFormFake";

interface Props {
  createComment: CreateCommentMutation;
  refreshSettings: FetchProp<typeof RefreshSettingsFetch>;
  sessionStorage: PromisifiedStorage;
  settings: PostCommentFormContainer_settings;
  viewer: PostCommentFormContainer_viewer | null;
  story: PostCommentFormContainer_story;
  showAuthPopup: ShowAuthPopupMutation;
  tab: COMMENTS_TAB;
  onChangeTab: (tab: COMMENTS_TAB) => void;
  commentsOrderBy?: COMMENT_SORT;
}

interface State {
  /** nudge will turn on the nudging behavior on the server */
  nudge: boolean;
  initialValues?: PropTypesOf<typeof PostCommentForm>["initialValues"];
  initialized: boolean;
  keepFormWhenClosed: boolean;
  submitStatus: SubmitStatus | null;
  notLoggedInDraft: string;
}

const contextKey = "postCommentFormBody";

/**
 * A temporary variable to save draft when user is not logged in.
 * This will be restored after the stream has refreshed.
 */
let preserveNotLoggedInDraft = "";

export class PostCommentFormContainer extends Component<Props, State> {
  public state: State = {
    initialized: false,
    nudge: true,
    keepFormWhenClosed:
      !!this.props.viewer &&
      !this.props.story.isClosed &&
      !this.props.settings.disableCommenting.enabled,
    submitStatus: null,
    notLoggedInDraft: "",
  };

  constructor(props: Props) {
    super(props);
    // Restore comment draft, if available.
    this.state.notLoggedInDraft = preserveNotLoggedInDraft;
    preserveNotLoggedInDraft = "";
    this.init();
  }

  public componentWillUnmount() {
    // Keep comment draft around. User might just have signed in and caused a reload.
    preserveNotLoggedInDraft = this.state.notLoggedInDraft;
  }

  private async init() {
    const body = await this.props.sessionStorage.getItem(contextKey);
    this.setState({
      initialValues: {
        body: body || this.state.notLoggedInDraft,
      },
    });
    this.setState({
      initialized: true,
    });
  }

  private disableNudge = () => {
    if (this.state.nudge) {
      this.setState({ nudge: false });
    }
  };

  private handleOnSubmit: PropTypesOf<
    typeof PostCommentForm
  >["onSubmit"] = async (input, form) => {
    try {
      if (this.props.tab === "FEATURED_COMMENTS") {
        this.props.onChangeTab("ALL_COMMENTS");
      }

      const submitStatus = getSubmitStatus(
        await this.props.createComment({
          storyID: this.props.story.id,
          nudge: this.state.nudge,
          commentsOrderBy: this.props.commentsOrderBy,
          ...input,
        })
      );
      if (submitStatus !== "RETRY") {
        form.initialize({});
      }
      this.setState({ submitStatus, nudge: true });
    } catch (error) {
      if (error instanceof InvalidRequestError) {
        if (shouldTriggerSettingsRefresh(error.code)) {
          await this.props.refreshSettings({ storyID: this.props.story.id });
        }
        return error.invalidArgs;
      }
      /**
       * Comment was caught in one of the moderation filters on the server.
       * We give the user another change to submit the comment, and we
       * turn off the nudging behavior on the next try.
       */
      if (error instanceof ModerationNudgeError) {
        this.disableNudge();
        return { [FORM_ERROR]: error.message };
      }
      // eslint-disable-next-line no-console
      console.error(error);
    }
    return;
  };

  private handleOnChange: PropTypesOf<typeof PostCommentForm>["onChange"] = (
    state,
    form
  ) => {
    if (this.state.submitStatus && state.dirty) {
      this.setState({ submitStatus: null });
    }
    if (state.values.body) {
      this.props.sessionStorage.setItem(contextKey, state.values.body);
    } else {
      this.props.sessionStorage.removeItem(contextKey);
    }
    // Reset errors whenever user clears the form.
    if (state.touched && state.touched.body && !state.values.body) {
      form.reset({});
    }
  };

  private handleDraftChange = (draft: string) => {
    this.setState({ notLoggedInDraft: draft });
  };

  private handleSignIn = () => this.props.showAuthPopup({ view: "SIGN_IN" });

  public render() {
    if (!this.state.initialized) {
      return null;
    }
    if (!this.state.keepFormWhenClosed) {
      if (this.props.settings.disableCommenting.enabled) {
        return (
          <PostCommentFormClosedSitewide
            story={this.props.story}
            message={this.props.settings.disableCommenting.message}
            showMessageBox={this.props.story.settings.messageBox.enabled}
          />
        );
      }
      if (this.props.story.isClosed) {
        return (
          <PostCommentFormClosed
            story={this.props.story}
            message={this.props.settings.closeCommenting.message}
            showMessageBox={this.props.story.settings.messageBox.enabled}
          />
        );
      }
    }
    if (!this.props.viewer) {
      return (
        <PostCommentFormFake
          draft={this.state.notLoggedInDraft}
          onDraftChange={this.handleDraftChange}
          story={this.props.story}
          showMessageBox={this.props.story.settings.messageBox.enabled}
          onSignIn={this.handleSignIn}
        />
      );
    }

    const scheduledForDeletion = Boolean(
      this.props.viewer.scheduledDeletionDate !== undefined &&
        this.props.viewer.scheduledDeletionDate !== null
    );

    return (
      <PostCommentForm
        story={this.props.story}
        onSubmit={this.handleOnSubmit}
        onChange={this.handleOnChange}
        initialValues={this.state.initialValues}
        min={
          (this.props.settings.charCount.enabled &&
            this.props.settings.charCount.min) ||
          null
        }
        max={
          (this.props.settings.charCount.enabled &&
            this.props.settings.charCount.max) ||
          null
        }
        disabled={
          this.props.settings.disableCommenting.enabled ||
          this.props.story.isClosed ||
          scheduledForDeletion
        }
        disabledMessage={
          (this.props.settings.disableCommenting.enabled &&
            this.props.settings.disableCommenting.message) ||
          (this.props.viewer.scheduledDeletionDate && (
            <Localized id="comments-postCommentForm-userScheduledForDeletion-warning">
              Commenting is disabled when your account is scheduled for
              deletion.
            </Localized>
          )) ||
          this.props.settings.closeCommenting.message
        }
        submitStatus={this.state.submitStatus}
        showMessageBox={this.props.story.settings.messageBox.enabled}
      />
    );
  }
}

const enhanced = withContext(({ sessionStorage }) => ({
  sessionStorage,
}))(
  withShowAuthPopupMutation(
    withCreateCommentMutation(
      withFetch(RefreshSettingsFetch)(
        withFragmentContainer<Props>({
          settings: graphql`
            fragment PostCommentFormContainer_settings on Settings {
              charCount {
                enabled
                min
                max
              }
              disableCommenting {
                enabled
                message
              }
              closeCommenting {
                message
              }
            }
          `,
          story: graphql`
            fragment PostCommentFormContainer_story on Story {
              id
              isClosed
              ...MessageBoxContainer_story
              settings {
                messageBox {
                  enabled
                }
              }
            }
          `,
          viewer: graphql`
            fragment PostCommentFormContainer_viewer on User {
              id
              scheduledDeletionDate
            }
          `,
        })(PostCommentFormContainer)
      )
    )
  )
);
export type PostCommentFormContainerProps = PropTypesOf<typeof enhanced>;
export default enhanced;
