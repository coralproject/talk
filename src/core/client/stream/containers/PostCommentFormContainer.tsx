import React, { Component, ReactNode } from "react";

import { BadUserInputError } from "talk-framework/lib/errors";
import { PropTypesOf } from "talk-framework/types";

import PostCommentForm, {
  PostCommentFormProps,
} from "../components/PostCommentForm";
import { CreateCommentMutation, withCreateCommentMutation } from "../mutations";

interface InnerProps {
  createComment: CreateCommentMutation;
  assetID: string;
  children?: ReactNode;
}

class PostCommentFormContainer extends Component<InnerProps> {
  private onSubmit: PostCommentFormProps["onSubmit"] = async (input, form) => {
    try {
      await this.props.createComment({
        assetID: this.props.assetID,
        ...input,
      });
      form.reset();
    } catch (error) {
      if (error instanceof BadUserInputError) {
        return error.invalidArgsLocalized;
      }
      // tslint:disable-next-line:no-console
      console.error(error);
    }
    return undefined;
  };
  public render() {
    return <PostCommentForm onSubmit={this.onSubmit} />;
  }
}

const enhanced = withCreateCommentMutation(PostCommentFormContainer);
export type PostCommentFormContainerProps = PropTypesOf<typeof enhanced>;
export default enhanced;
