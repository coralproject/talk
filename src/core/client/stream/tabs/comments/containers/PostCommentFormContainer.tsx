import React, { Component } from "react";

import { withContext } from "talk-framework/lib/bootstrap";
import { BadUserInputError } from "talk-framework/lib/errors";
import { PromisifiedStorage } from "talk-framework/lib/storage";
import { PropTypesOf } from "talk-framework/types";

import {
  CreateCommentMutation,
  withCreateCommentMutation,
} from "talk-stream/mutations";
import PostCommentForm, {
  PostCommentFormProps,
} from "../components/PostCommentForm";

interface InnerProps {
  createComment: CreateCommentMutation;
  storyID: string;
  sessionStorage: PromisifiedStorage;
}

interface State {
  initialValues?: PostCommentFormProps["initialValues"];
  initialized: boolean;
}

const contextKey = "postCommentFormBody";

export class PostCommentFormContainer extends Component<InnerProps, State> {
  public state: State = { initialized: false };

  constructor(props: InnerProps) {
    super(props);
    this.init();
  }

  private async init() {
    const body = await this.props.sessionStorage.getItem(contextKey);
    if (body) {
      this.setState({
        initialValues: {
          body,
        },
      });
    }
    this.setState({
      initialized: true,
    });
  }

  private handleOnSubmit: PostCommentFormProps["onSubmit"] = async (
    input,
    form
  ) => {
    try {
      await this.props.createComment({
        storyID: this.props.storyID,
        ...input,
      });
      form.reset({});
    } catch (error) {
      if (error instanceof BadUserInputError) {
        return error.invalidArgsLocalized;
      }
      // tslint:disable-next-line:no-console
      console.error(error);
    }
    return undefined;
  };

  private handleOnChange: PostCommentFormProps["onChange"] = state => {
    if (state.values.body) {
      this.props.sessionStorage.setItem(contextKey, state.values.body);
    } else {
      this.props.sessionStorage.removeItem(contextKey);
    }
  };

  public render() {
    if (!this.state.initialized) {
      return null;
    }
    return (
      <PostCommentForm
        onSubmit={this.handleOnSubmit}
        onChange={this.handleOnChange}
        initialValues={this.state.initialValues}
      />
    );
  }
}

const enhanced = withContext(({ sessionStorage }) => ({
  sessionStorage,
}))(withCreateCommentMutation(PostCommentFormContainer));
export type PostCommentFormContainerProps = PropTypesOf<typeof enhanced>;
export default enhanced;
