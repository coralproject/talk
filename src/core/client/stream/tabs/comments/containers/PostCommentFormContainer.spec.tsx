import { shallow } from "enzyme";
import { noop } from "lodash";
import React from "react";
import sinon from "sinon";

import { timeout } from "talk-common/utils";
import { createPromisifiedStorage } from "talk-framework/lib/storage";
import { removeFragmentRefs } from "talk-framework/testHelpers";
import { PropTypesOf } from "talk-framework/types";

import { PostCommentFormContainer } from "./PostCommentFormContainer";

const contextKey = "postCommentFormBody";
const PostCommentFormContainerN = removeFragmentRefs(PostCommentFormContainer);

function createDefaultProps(): PropTypesOf<typeof PostCommentFormContainerN> {
  return {
    createComment: noop as any,
    refreshSettings: noop as any,
    storyID: "story-id",
    sessionStorage: createPromisifiedStorage(),
    settings: {
      charCount: {
        enabled: true,
        min: 3,
        max: 100,
      },
    },
  };
}

it("renders correctly", async () => {
  const props: PropTypesOf<typeof PostCommentFormContainerN> = {
    ...createDefaultProps(),
  };

  const wrapper = shallow(<PostCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it("renders with initialValues", async () => {
  const props: PropTypesOf<typeof PostCommentFormContainerN> = {
    ...createDefaultProps(),
  };

  await props.sessionStorage.setItem(contextKey, "Hello World!");

  const wrapper = shallow(<PostCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it("save values", async () => {
  const props: PropTypesOf<typeof PostCommentFormContainerN> = {
    ...createDefaultProps(),
  };

  await props.sessionStorage.setItem(contextKey, "Hello World!");

  const wrapper = shallow(<PostCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  wrapper
    .first()
    .props()
    .onChange({ values: { body: "changed" } });
  expect(await props.sessionStorage.getItem(contextKey)).toBe("changed");
});

it("creates a comment", async () => {
  const storyID = "story-id";
  const input = { body: "Hello World!" };
  const createCommentStub = sinon.stub();
  const form = { reset: noop };
  const formMock = sinon.mock(form);
  formMock
    .expects("reset")
    .withArgs({})
    .once();

  const props: PropTypesOf<typeof PostCommentFormContainerN> = {
    ...createDefaultProps(),
    createComment: createCommentStub,
    storyID,
  };

  await props.sessionStorage.setItem(contextKey, "Hello World!");

  const wrapper = shallow(<PostCommentFormContainerN {...props} />);
  await timeout();
  wrapper.update();
  wrapper
    .first()
    .props()
    .onSubmit(input, form);
  expect(
    createCommentStub.calledWith({
      storyID,
      ...input,
    })
  ).toBeTruthy();
  await timeout();
  formMock.verify();
});
