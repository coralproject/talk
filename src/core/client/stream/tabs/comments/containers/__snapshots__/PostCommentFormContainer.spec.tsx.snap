// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<PostCommentForm
  max={100}
  min={3}
  onChange={[Function]}
  onSubmit={[Function]}
/>
`;

exports[`renders with initialValues 1`] = `
<PostCommentForm
  initialValues={
    Object {
      "body": "Hello World!",
    }
  }
  max={100}
  min={3}
  onChange={[Function]}
  onSubmit={[Function]}
/>
`;
